client/parser/format.go

```diff
@@ -0,0 +1,748 @@
+		texts:       []string{},
+		out:         out,
+	}
+
+	for _, opt := range opts {
+		opt(formatter)
+	}
+
+	return formatter
+}
+
+// line constructs and returns the current line being formatted.
+func (f *formatter) line() string {
+	text := strings.Join(f.texts, " ")
+	if len(text) == 0 {
+		return text
+	} else {
+		indent := strings.Repeat(f.indentText, f.indentLevel)
+		return indent + text
+	}
+}
+
+// emptyLine writes an empty line
+func (f *formatter) emptyLine() {
+	if f.varDef == None {
+		f.flush()
+		f.out.Write([]byte("\n"))
+	}
+}
+
+// flush flushes out the current line to the output.
+func (f *formatter) flush() {
+	if len(f.texts) > 0 && f.varDef == None {
+		f.out.Write([]byte(f.line() + "\n"))
+		f.texts = []string{}
+	}
+}
+
+// indent increments the indentation level of subsequent formatting.
+func (f *formatter) indent() {
+	switch f.varDef {
+	case LastNode:
+		// During the last node of a var def, we are allowed to effectively exit
+		// the var def and continue formatting on the next line.
+		f.varDef = None
+		fallthrough
+	case None:
+		f.flush()
+		f.indentLevel++
+	}
+}
+
+// unindent decrements the indentation level of subsequent formatting.
+// A corresponding unindent should always be called after calling indent.
+func (f *formatter) unindent() {
+	if f.varDef == None {
+		f.flush()
+		f.indentLevel--
+	}
+}
+
+// write formats text to the output with indentation, wrapping, and spacing.
+// Each "text" is an unwrappable token, i.e. wrapping only happens between text.
+func (f *formatter) write(text string) {
+	f.texts = append(f.texts, text)
+	if len(f.line()) > f.softWrapLen && f.varDef == None {
+		f.texts = f.texts[0 : len(f.texts)-1]
+		f.flush()
+		f.texts = append(f.texts, text)
+	}
+}
+
+// formatWithDuration handles duration formatting.
+// Durations are formatted with possible text directly pre/post (no spaces),
+// i.e. note pitches preceding durations.
+// All durations are treated as a single unwrappable text with the exception of
+// barlines which cause a duration to be split into separate texts.
+func (f *formatter) formatWithDuration(
+	pre string, duration ASTNode, post string,
+) error {
+	text := strings.Builder{}
+	text.WriteString(pre)
+	shouldTie := false
+
+	for i, child := range duration.Children {
+		switch child.Type {
+
+		default:
+			return fmt.Errorf(
+				"unexpected DurationNode %#v during formatting", child,
+			)
+
+		case BarlineNode:
+			if i == len(duration.Children)-1 {
+				// The final duration is a barline
+				// We write out any post text before the barline for clarity
+				text.WriteString(post)
+			}
+
+			// Barlines in a duration split formatting into separate texts
+			if text.Len() > 0 {
+				f.write(text.String())
+			}
+			f.write("|")
+
+			text.Reset()
+
+		case NoteLengthMsNode:
+			if shouldTie {
+				text.WriteString("~")
+			}
+
+			// NoteLengthMs is parsed to us as ms, but can be coded using s/ms
+			// We will try our best to use seconds first
```

Comment:
Wouldn't this result in a potentially different AST than the original code? It seems like to keep things simple, it would be better if we just formatted this as milliseconds.

Is the issue here that the parser normalizes seconds into a NoteLengthMsNode, so at this point, we don't know if the original code was seconds or milliseconds? If so, I don't think it would be a problem to add a NoteLengthSecondsNode, so that we can preserve the original intention.

Reply:
Ah I thought I was doing something clever, but this is just an oversight from me.

So yeah the scanner just produces the literal as a ms float64. So "1s" turns into "1000". 

I will make the change to add a `NoteLengthSecondsNode` and equivalents then. So we can preserve the original intention.

Reply:
Made changes here: https://github.com/alda-lang/alda/pull/474/commits/902938c74bb72bb12fe2b303ab7497acc496d65f.

I essentially added 
- `NoteLengthSeconds` token to the scanner
- `NoteLengthSecondsNode` node to the parser
- `model.NoteLengthSeconds` score update to model

Leaving to let you resolve.